### PR TITLE
feat: Native Merkle Trees as a native NodeJS module

### PR DESCRIPTION
**Pull request by alexghr**
The document proposes a new interface to the native merkle tree implementation from #3 using Node-API and `bindings` instead of message passing.

### CODE REVIEW

File: in-progress/5040-native-merkle-trees-napi.md
Comment (PhilWindle) [inline in-progress/5040-native-merkle-trees-napi.md:14]: I don't think we ever intended to spawn a new process every time. IT was always going to be a long-lived process.

Comment (alexghr) [reply in-progress/5040-native-merkle-trees-napi.md:14]: I itended to remove this paragraph but missed before pushing the commit :facepalm:

File: in-progress/5040-native-merkle-trees-napi.md
Comment (spalladino) [inline in-progress/5040-native-merkle-trees-napi.md:262]: I recall @ludamad mentioned this was an issue when generating WASM. I assume it is not a problem in this context, right?

File: in-progress/5040-native-merkle-trees-napi.md
Comment (spalladino) [inline in-progress/5040-native-merkle-trees-napi.md:107]: Nit: where is `callback` used?

File: in-progress/5040-native-merkle-trees-napi.md
Comment (spalladino) [inline in-progress/5040-native-merkle-trees-napi.md:160]: If we implemented this, does it make sense to keep maintaining the message passing interface? Feels like it'd be simpler to deprecate that one, and just have one exported function for each actual tree function, right? No strong feelings here though.

**Review by spalladino — APPROVED** [review]
Thanks for looking so thoroughly into this! Looks great to me, but given the amount of cpp in here I don't think I'm the right person for approving. I'd try to get one of @charlielye or @ludamad here if possible.

File: in-progress/5040-native-merkle-trees-napi.md
Comment (fcarreiro) [inline in-progress/5040-native-merkle-trees-napi.md:1]: Nice alternative and investigation!

File: in-progress/5040-native-merkle-trees-napi.md
Comment (alexghr) [reply in-progress/5040-native-merkle-trees-napi.md:107]: Fixed :grimacing:

File: in-progress/5040-native-merkle-trees-napi.md
Comment (alexghr) [reply in-progress/5040-native-merkle-trees-napi.md:179]: I think it'll come down to the kind of async ops that we'll run inside `AsyncWorker` instances and how we design the code (someone with more C++ experience could chime in here). It might be easier to refactor the generic `tree_op_callback` to return a msgpacked value (currently in this spec the callback must return a `bb::fr`) as then it becomes trivial to send it back to JS as a buffer vs serializing to POJOs in the `OnOk` callback.

Comment (just-mitch) [reply in-progress/5040-native-merkle-trees-napi.md:179]: If we kept the message passing interface, then wouldn't we not need to worry about keeping types in parity across TS and cpp? And we wouldn't need to deserialize `any` in TS? And minor benefit that other clients of bb (if/when that day should come) that want to use these trees without the nodeapi could more easily integrate them?

File: in-progress/5040-native-merkle-trees-napi.md
Comment (just-mitch) [inline in-progress/5040-native-merkle-trees-napi.md:111]: Just for understanding, this should `deferred.Resolve(leaf)` right?

**Review by just-mitch — APPROVED** [review]
Looks good to me! I'm still a fan of a common message interface, even if this nodeapi wraps it, so I'd weigh the pros/cons there, but generally I like this a lot!

**Comment by just-mitch** [comment]
@alexghr Are we still waiting on people's explicit approval? Or has this entered implementation? Should we merge the design?

**Pull request merged**
